### PR TITLE
Rewrite VS pipelines to use chunks

### PR DESCRIPTION
**Pull request by mwojcikowski**
VS pipelines will move `chunksize` of molecules at a time. The idea is to lower cost of pickling objects for every molecule. Finally we should have at best `n_cpus * chunksize` molecules in memory.
Missing bits:

- [x] make executing loop in `fetch()` method lazy
- [x] ensure that there is max `n_cpus * chunksize` molecules in memory
- [x] introduce chunksize to `oddt_cli`

### CODE REVIEW

File: oddt/utils.py [hunk 1]
```diff
@@ -64,3 +65,25 @@ def check_molecule(mol,
 
     if non_zero_atoms and len(mol.atoms) == 0:
         raise ValueError('Molecule "%s" has zero atoms.' % mol.title)
+
+
+def compose_iter(iterable, funcs):
```
Comment (marta-sd) [inline oddt/utils.py]: Consider using `*funcs` instead. I prefer `fun(a, f1, f2, f3)` than `fun(a, (f1, f2, f2))`.

File: oddt/utils.py [hunk 2]
```diff
@@ -64,3 +65,25 @@ def check_molecule(mol,
...
+    """Chain functions and apply them to iterable, by exhausting the iterable.
+    Functions are executed in the order from funcs."""
+    for func in funcs:
+        iterable = func(iterable)
+    return list(iterable)
+
+
+def chunker(iterable, chunksize=100):
+    """Generate chunks from iterable"""
```
Comment (marta-sd) [inline oddt/utils.py]: You should change the docstring, the argument must be an iterator.

File: oddt/virtualscreening.py [hunk 3]
```diff
@@ -92,65 +136,35 @@ def apply_filter(self, expression, soft_fail=0):
...
+                                           expression=['mol.molwt < 300',
+                                                       'mol.HBA1 <= 3',
+                                                       'mol.HBD <= 3',
+                                                       'mol.logP <= 3'],
+                                           soft_fail=soft_fail)))
             # PAINS filter
             elif expression.lower() in ['pains']:
                 pains_smarts = {}
-                with open(dirname(__file__)+'/filter/pains.smarts') as pains_file:
+                with open(dirname(__file__) + '/filter/pains.smarts') as pains_file:
```
Comment (marta-sd) [inline oddt/virtualscreening.py]: It's not platform independent

File: oddt/virtualscreening.py [hunk 4]
```diff
@@ -92,65 +136,35 @@ def apply_filter(self, expression, soft_fail=0):
...
-                                           'mol.logP <= 5'],
-                                          soft_fail=soft_fail)
+                self._pipe.append((partial(_filter,
+                                           expression=['mol.molwt < 500',
+                                                       'mol.HBA1 <= 10',
+                                                       'mol.HBD <= 5',
+                                                       'mol.logP <= 5'],
+                                           soft_fail=soft_fail)))
             # Rule of three
             elif expression.lower() in ['ro3']:
```
Comment (marta-sd) [inline oddt/virtualscreening.py]: `expression.lower() == 'ro3'`

File: oddt/virtualscreening.py [hunk 5]
```diff
@@ -92,65 +136,35 @@ def apply_filter(self, expression, soft_fail=0):
...
-                                           'mol.logP <= 3'],
-                                          soft_fail=soft_fail)
+                self._pipe.append((partial(_filter,
+                                           expression=['mol.molwt < 300',
+                                                       'mol.HBA1 <= 3',
+                                                       'mol.HBD <= 3',
+                                                       'mol.logP <= 3'],
+                                           soft_fail=soft_fail)))
             # PAINS filter
             elif expression.lower() in ['pains']:
```
Comment (marta-sd) [inline oddt/virtualscreening.py]: `expression.lower() == 'pains'`

File: oddt/virtualscreening.py [hunk 6]
```diff
@@ -204,13 +218,12 @@ def similarity(self, method, query, cutoff=0.9, protein=None):
...
         else:
             raise ValueError('Similarity filter "%s" is not supported.' % method)
-        query_fps = [(gen(q) if protein is None else gen(q, protein))
-                     for q in query]
-        self._pipe = filter(lambda q: any(dist(gen(q) if protein is None
-                                               else gen(q, protein),
-                                               q_fp) >= float(cutoff)
-                                          for q_fp in query_fps),
-                            self._pipe)
+        query_fps = [gen(q) for q in query]
```
Comment (marta-sd) [inline oddt/virtualscreening.py]: Add a comments here and in L224, smth like "generate FPs for query molecules" + "use the same FP generator for molecules in pipe".

File: oddt/utils.py (hunk 1 above)
Comment (mwojcikowski) [reply oddt/utils.py]: If we did we couldn't make a partial of it without first positional argument, like now.

```python
composed_f1_f2_f3 = partial(compose_iter, funcs=[f1, f2, f3])
composed_f1_f2_f3(range(3))
```

**Pull request merged**